workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,238 @@
+                return;
+            }
+
+            // Mark reader components ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = componentIdToReaderOrWriterSpatialOSBehaviours[componentId];
+            foreach (var spatialOSBehaviour in relevantReaderSpatialOSBehaviours)
+            {
+                MarkComponentReady(spatialOSBehaviour);
+            }
         }
 
         public void RemoveComponent(uint componentId)
         {
-            // Mark readers as to be disabled
+            var writerWanted = componentIdToWriterSpatialOSBehaviours.ContainsKey(componentId);
+            var readerWanted = !writerWanted && componentIdToReaderOrWriterSpatialOSBehaviours.ContainsKey(componentId);
+            if (!readerWanted && !writerWanted)
+            {
+                LogNotInterestedError(componentId);
+                return;
+            }
+
+            // Mark reader components not ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = componentIdToReaderOrWriterSpatialOSBehaviours[componentId];
+            foreach (var spatialOSBehaviour in relevantReaderSpatialOSBehaviours)
+            {
+                MarkComponentNotReady(spatialOSBehaviour);
+            }
         }
 
         public void ChangeAuthority(uint componentId, Authority authority)
         {
-            // Mark writers as to be enabled/disabled
+            var writerWanted = componentIdToWriterSpatialOSBehaviours.ContainsKey(componentId);
+            var readerWanted = !writerWanted && componentIdToReaderOrWriterSpatialOSBehaviours.ContainsKey(componentId);
+            if (!readerWanted && !writerWanted)
+            {
+                LogNotInterestedError(componentId);
+                return;
+            }
+
+            if (authority == Authority.Authoritative)
+            {
+                // Mark writer components ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = componentIdToWriterSpatialOSBehaviours[componentId];
+                foreach (var spatialOSBehaviour in relevantWriterSpatialOSBehaviours)
+                {
+                    MarkComponentReady(spatialOSBehaviour);
+                }
+            }
+            else if (authority == Authority.NotAuthoritative)
+            {
+                // Mark writer components not ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = componentIdToWriterSpatialOSBehaviours[componentId];
+                foreach (var spatialOSBehaviour in relevantWriterSpatialOSBehaviours)
+                {
+                    MarkComponentNotReady(spatialOSBehaviour);
+                }
+            }
+        }
+
+        private void MarkComponentReady(MonoBehaviour spatialOSBehaviour)
+        {
+            // Inject all Readers/Writers at once when all requirements are met
+            if (--numUnSatisfiedReadersWriters[spatialOSBehaviour] == 0)
```

Comment:
I'm not sure that this will set the value back - also from a legibility POV it would be nice to split this into multiple lines

Reply:
I'll check the functionality (I'm pretty sure it should work though, given that it compiled at all) and split it

Reply:
Yeah it will compile - but as far as I understand, it will read the value in the dictionary, subtract, and do the comparison. 

But it won't set the new value back in the dictionary - meaning that if there are 2 reader or writers that condition can never be true 😉